### ISSUE TITLE
Units

### ISSUE THREAD

**Issue by rocketman768** [beerjson/beerjson#16]
Hi all,

I didn't follow much the development of BeerXML2, so I apologize if I'm dragging out old topics. I am curious about the allowed units. What is the use case for allowing so many different units inside the JSON itself?

I think it could make this standard simpler (most of measureable_units.json could be gone) if each value were expressed in a single unit system, like [meter-kilogram-seconds](https://en.wikipedia.org/wiki/MKS_system_of_units). Parsing would be simpler for apps as well, since they wouldn't have to support so many possible conversions. I know the first thing we would do in Brewtarget is just to run all this through our complex unit conversion code, and make sure we support all the units in the spec., to get everything into MKS internally.

I understand that not everything (like color) will fit into a unit system, but I think the things that do fit squarely (time, temperature, mass, volume,...) could be very simplified. What do you think?

Thank you!
-Philip

**Comment by superroma** [comment]
Well, the [latest development of BeerXML 2](http://users.speakeasy.net/%7Eantonw/beer_xml/BeerXML_v2_01.pdf) defines multiple units for every type of measurment.
Please note that I'm not a pro brewer, so I cannot really tell if it is good or bad. As sofware developer, I agree, one system is better, we do convert to metric in our calculator, but probably for human reading JSON it is not that good.

In the end, I'm hoping to have a working group, that would decide this, I personally will be happy with metric system.

**Comment by krutilin** [comment]
Yes, there are set of questions to the "measureable_units.json" schema. And looks like it could be refactored and simplified.
Not sure about the only metric system.

But at the moment we have some duplication in jsons with data. For example:
```json
        "grain_temperature": {
          "degrees": "C",
          "value": 22.0
        },
        "infuse_amount": {
          "volume": "l",
          "value": 10.0
        }
```
could be simplified to 
```json
        "grain_temperature": {
          "scale": "C",
          "value": 22.0
        },
        "infuse_amount": {
          "scale": "l",
          "value": 10.0
        }
```
without pointing that it is "degrees" or "volume", because we could understand and parse it according to unit string. I see that in brewtarget you also use "scale" and "value" structures. And other guys like https://www.npmjs.com/package/convert-units do the same. We should prepare some kind of spike with tests and give it to the community review.

**Comment by superroma** [comment]
"units" instead of "scale" sounds better to me

**Comment by jdominiczak** [comment]
@superroma I agree with your desire to have everything in a single unit structure, but I do see value in supporting multiple unit schemes.  From a readability aspect, I'd like to view my recipes in the units of my choosing instead of having to use software to convert them.  

@krutilin I like the structure you're proposing but I do agree with @superroma that units is a more accurate description.  What you do loose is the validation of appropriate types of units (eg. is a temperature property being defined by a temperature unit instead of a weight/volume/gravity unit.

**Comment by pricelessbrewing** [comment]
I would prefer non group unit systems, as it's not uncommon to mix unit
systems like using hop weight as grams, or yeast and starters as mL if it
was rigidly fixed as either metric or imperial there would be a large
number of users that didn't fit that system.

**Comment by krutilin** [comment]
I did schema refactoring for units validation in following way:
```json
        "grain_temperature": {
          "units": "C",
          "value": 22.0
        }
```
Also set "units" and "value" fields required for validation (it was possible to make an empty object). And fixed tests according to these changes. All stuff in the following branch https://github.com/beerjson/beerjson/tree/units-schema-refactoring

@rocketman768 @jdominiczak @pricelessbrewing

**Comment by superroma** [comment]
It looks clean and readable. But it is readable because there is a word "temperature" in the property names. Some cases are not that clear: yeast addition and grain addition both has "amount" property of the different type:
```

{
              "name": "Pale Malt(2-row)",
              "type": "grain",
              "color": {
                "units": "L",
                "value": 1.8
              },
              "origin": "America",
              "amount": {
                "units": "lb",
                "value": 8.0
              }
            },
"yeast_additions": [
            {
              "name": "English Ale",
              "type": "ale",
              "form": "liquid",
              "laboratory": "Ithaca Beer C",
              "product_id": "",
              "amount": {
                "units": "ml",
                "value": 10.0
              }

```
So the only way for reader to guess that this is volume - is to look at "ml" unit.
What about having actual unit name instead of value?
```
       "grain_temperature": {
          "units": "C",
          "temperature": 22.0
        }
...
              "amount": {
                "units": "ml",
                "volume": 0.0
              }
...
              "amount": {
                "units": "lb",
                "weight": 8.0
              }

```
This look more readable. Parsing code will have to deal with each unit type anyway, so there is not much harm

**Comment by superroma** [comment]
@rocketman768 @jdominiczak @pricelessbrewing - what do you guys think?

**Comment by jdominiczak** [comment]
I would agree with you @superroma.  This would force the correct type of unit.  I like what you have proposed.

**Comment by rocketman768** [comment]
@superroma it seems from the discussion that you are expecting a significant percentage of the users to edit these files by hand. If that is the case, it looks OK.

**Comment by superroma** [comment]
@rocketman768, users as those who use BeerJSON as exchange format - no, i don't expect them to edit these files. 
Users as developers who deal with import/export issues, bugs, errors, format differences - yes, I expect them to at least read those files. 
After all, if we cannot stick to the single unit system, we'll have all those unit conversion logic in the code, so it should not add much code to deal with different keys for different unit types.

**Comment by mrichar1** [comment]
Hi - been meaning to get into this project a bit more after seeing it a while ago, so thought I'd start with some comments on issues - apologies if I'm repeating other's points, or missing fundamentals that have already been decided!

I'd suggest that within BeerJSON units should ultimately be free-form text. Trying to store only in a single unit scheme (e.g. SI) means that conversions need to be in place before the project becomes useful. We already see at least 3 common schemes in use (metric, US imperial, UK imperial) and you'd need to have full conversion set up between them. And trying to mandate a definitive list will only result in corner cases where someone couldn't add their recipe (I've seen some fantastic historical recipes with arcane measures, like "a weight of herbs equal to that of a duck's egg").

The downsides of free-form text can be mitigated by using other libraries that already provide lists of units, and their conversions - such as https://github.com/ben-ng/convert-units/tree/master/lib/definitions.Using these, rather than reinventing the wheel could save a lot of work. (It would even be possible to mirror and JSONify their definitions if it was felt that having that information 'internally' was important). 

Oh, and on a simpler note - please can there be a way to specify a count of something - like '5 apples' :smiley:

**Comment by pricelessbrewing** [comment]
There's also the unit type pH missing. 

In the documentation it should be listed as 

Acidity Unit Type

pH 

 Example:  <start pH>5.2</start pH>

or 

 Example:  <start acidity scale="pH"> 5.2 </start acidity>

@mrichar1 While that seems nice, it also is quite cumbersome with minimal usefulness IMO.

**Comment by mrichar1** [comment]
@pricelessbrewing - can you explain why you feel it would be cumbersome and minimally useful in comparison to the issues I raised with the other 2 alternative approaches?

**Comment by pricelessbrewing** [comment]
The link you posted above is broken, fyi.

I was specifically referring to the "each" count, sorry I wasn't more specific. The main problem I see with it is that it would require automatic conversion into a weight or volume unit in order to be compatible with _required_ variables such as potential/yield for a fermentable, which is defined at a per unit weight per unit volume basis. Could probably find an api for that too if needed, but I think that's more a problem for the brewing software to deal with than the beerJSON standard to support. 

I completely agree with the free form units, and not a specific unit system (SI / ISO vs US imperial etc).

**Comment by mrichar1** [comment]
Ahh, good - nice to know you agree on the main point about free form units (I did wonder that if you were rejecting it your response was a bit brief! :smiley: )

Yes, I can see the issue with 'each' in that regard - I wonder if perhaps it is something that might only make sense for adjuncts or other ingredients where there is no expectation of it having an effect  on volume, yield etc? (Though I haven't looked into the schema enough to know if this is a plausible idea).

I'm mostly thinking of lots of old brewing books which specify recipes with 'each' ingredients where a weight would either be less useful (1 clove vs 0.28g cloves) or just hard to find retrospectively (I can't go weigh 12 elderflower heads until they're in bloom again).

**Comment by rmsrosa** [comment]
Hi, I only recently found out about this project, but I am very much interested on it, as I wrote in the chat. I like the way things are going. If you don't mind me saying it, I like the idea of allowing different unit systems (a tbsp of coriander seeds, etc.). 

I would rather use "value" regardless of the unit type, instead of specifying it as "weight", "volume", etc. The unit type would be implicit in the object and I don't mind it.

I don't understand why you use "units" instead of "unit", when specifying "one" of the possible units for a given quantity.

I like "units" instead of "measurable_units", and it wouldn't conflict with "unit", if we drop the plural as mentioned in the previous paragraph.

Along the line of what @mrichar1 said, I thing we could allow, in a miscellaneous record, a key "amount_each", complementing "amount" and "amount_as_weight", with the type being a floating point, so a "0.5 scorpion pepper" would be acceptable. But that is not a big deal. We could just guesstimate the weight and write "1/2 scorpion pepper" in the "notes" field. Similarly for "1 1/2" vials of White Labs yeast or "0.5 each". Just as a side note, I use BeerTools Pro and, for the yeast addition, it allows explicitly "weight", "volume" and "each" as possible units (although no conversion is used when changing to "each").

**Comment by superroma** [comment]
>I would rather use "value" regardless of the unit type, instead of specifying it as "weight", "volume", etc. The unit type would be implicit in the object and I don't mind it.

I thought this way JS would be more human-readable. For code it makes not much difference (i guess), since you'll need to write some unit conversion function there anyway.

> I don't understand why you use "units" instead of "unit", when specifying "one" of the possible units for a given quantity.

I'm not a native English speaker, so fixes and corrections are welcome!

> I like "units" instead of "measurable_units", and it wouldn't conflict with "unit", if we drop the plural as mentioned in the previous paragraph.

Can be changed too - this is what it was called in the latest BeerXML 2 draft from 2007. But please note, this is just a name for the schema - there will be no "measurable_units" string anywhere in actual recipe JSON files.

**Comment by rmsrosa** [comment]
>> I like "units" instead of "measurable_units", and it wouldn't conflict with "unit", if we drop the plural as mentioned in the previous paragraph.

> Can be changed too - this is what it was called in the latest BeerXML 2 draft from 2007. But please > note, this is just a name for the schema - there will be no "measurable_units" string anywhere in actual recipe JSON files.

I see. I thought it would appear as a key name somewhere, but that is right. Then I guess we shouldn't bother with that.

**Comment by pricelessbrewing** [comment]
Listing all the unit types I can think of that are not present currently. 

Current Density unit type should change to SpecificDensity. 

Should Density be created (g/L, lb/gallon etc ). Or if there's a way to do something like MassUnitType/VolumeUnitType that would be good too. Example quantity being measured, hopping rate (12 lbs per barrel), titratable acidity (g/L), 

Specific volume should be created. Example quantity being measured, mash thickness. This is the inverse of density, qt/lb, L/kg, gal/lb etc. 

Should Distance be re-created? Example quantity being measured, equipment dimensions, some software allows for volumetric measurements to be estimated by measuring depth of the liquid (how tall is the wort in your kettle). Would be nice to support this natively, instead of them adding it internally then removing it when exporting to other software, although the explicit volume should always be the preferred quantity, and required. 

Should concentration be created (ppm, ppb)? Would be nice to eventually support things like diacetyl concentration, titratable acidity, ester concentration, for those with labs. This is obviously more professionally oriented, but would likely go a long way to getting support from pros.

**Comment by rmsrosa** [comment]
Good point. What BeerXML and the current beerjson call Density is not really density. In my software, I call it Extract. I suggest to use the same. Extract with units Plato or specific gravity. Specific gravity is, well, too specific.

And I agree that we should also  have a quantity Density with its real meaning.

I don't know what do you mean about re-creating Distance. That is Length. 

I think Concentration should be created, as well.

**Comment by rmsrosa** [comment]
On the same topic, we have the same letter "L" being used for Liter, Lintner and Lovibond. Actually I think Lintner is not use here, but anyway. It is not a big deal, because they are used for different quantities, but I think it would be nicer to, say, use Lov for Lovibond. 

By the way, on googling  about this, I stumbled on the page 

https://code.google.com/archive/p/beerson/wikis/NameValuePairs.wiki

which uses LOV for Lovibond (but since it is not an acronym, I think we should not use all capitals).

This page also aims towards a json replacement for BeerXML, but I think it is closer to BeerXML 1.0 (I didn't check carefully), and it does not have a schema.

Someone had already mentioned a yet another project http://www.beerjson.com/contact/, which apparently didn't take off.

Anyway, I just thought I should mention that. The question is about L or Lov for Lovibond.

**Comment by pricelessbrewing** [comment]
> Length vs distance

Thanks, looks like I overlooked that. I thought for sure that it was deleted in a previous pull request. 

>Lintner, Lovibond, Liter

Liter, I don't think that's a problem. 

I would propose we use the nomenclature, and units used in malt lot analysis sheets, which is common in homebrew use but not quite as "official". 

ºLovibond for lovibond, and L for lintner. We could use the full word if it increases the contrast between them, which is probably a good idea anyways. 

`Extract vs Specific Density`

The problem I can think of with extract, is that traditionally extract is expressed in a mass unit (kg/lb). Where it's the total mass of sugar dissolved in the wort. Which is not part of the schema as it currently stands, but is part of Beersmith, and is used in more traditional brewing cultures (mostly german, and old breweries that haven't switched to more modern measurements). Search keyword if original/total extract. 

Specific gravity / density is the quantity being measured here, but even though sg and plato are roughly equivalent they are technically very different quantities. SG is the ratio of densities in relation to water at STP, while plato is the percent of extract (total dry sugar mass) to the total weight of wort (water weight + sugar weight). Technically this means that both are unitless ratios, or percentages, which is kinda funny to me. 

I'd be fine with extract instead of density, although I think specific density or specific gravity would be more appropriate, just noting potential issues.

**Comment by rmsrosa** [comment]
I think we need an abbreviation for Lovibond; using °Lovibond might be too large to be displayed in some softwares. Of course the software may use a different abbreviation, but it would be nice to have the same one. I still prefer Lov to make it clear, but if that is unusual and the norm is to use °L, then I think L is better than no abbreviation. Fortunately, Lintner doesn't need to be used since it is the only unit for diastatic power.

I don't have this dilemma with Extract and Specific Density being different physical measures. The theoretical and practical applications are mixed. We measure specific density to estimate the extract, or sugar content, which is what is relevant for the fermentation, since the yeast/microbe eats the sugar, not the density. And it is the extract that goes into the formula for calculating ABV, Calories, etc. At the same type, EBC, SRM and Lovibond are not different units of the same thing, they are actually slightly different ways to infer a one-dimensional representation of the color of the beer. IBU is also sometimes considered as an indirect measure of mg/L of iso-alpha acids, although the story here seems to be more obscure. So, anyway, I think it is fine to call the quantity as "Extract" and to have Plato and specific gravity as two possible units. I like that better.

**Comment by rmsrosa** [comment]
I was thinking about branching and working on this issue, but I am not sure we agreed on a final form. 

One issue was about whether or not use "mass" and "volume" explicitly instead of "value", or not. Another suggestion was to include "each" and other measures. I see another way of achieving these things: 

"Amount" is not clarifying, and, depending on the context, it can mean "mass" or "volume". So, my suggestions is to replace "amount", "amount_as_weight", and "amount_as_volume" explicitly by "mass" and "volume", and then, inside the object, we would just have "unit" and "value". 

In some cases, for instance in misc.json, we can add an amount "each", for a countable addition. In yeast.json, we can also add "each", to represent a vial of White Labs or a smack-pack of Wyeast, or a sachet of dry yeast, etc. Whether it is a vial or a smack-pack or something else would be described by the title of the product itself, and then the amount would just be "each". Of course, the software should also give the option for the user to explicitly describe the amount in volume or mass, if desired.

What do you think?

**Comment by krutilin** [comment]
I made some attempts to store my recipes from paper notebook to beerjson and found that it is not convenient to write units in this way:

```       
 "amount": {
                "units": "ml",
                "weight": 8.0
              }
```
I always forgot what is correct "weight" or "mass" and etc. So I agree that better to store knowlege about mass or volume in "amount_as_weight", and "amount_as_volume" constructions:
```       
 "amount_as_weight": {
                "units": "ml",
                "value": 8.0
              }
```
as for software it does not matter, it will parse any json. But for human readable recipe I like to have units\value construction. Because it does not force you to remember additional info each time.

**Comment by rmsrosa** [comment]
Yes, I was proposing to write
```
 "volume": {
                "unit": "ml",
                "value": 8.0
              }
  "mass": {
                "unit": "g",
                "value": 8.0
              }
```
and, for the dimensionless quantities, we could use either
```
  "number": {
                "unit": "dimensionless",
                "value": 8.0
              }
```
or simply
```
  "number" : 8.0
```
The first option for the dimensionless quantity is based on http://www.iso.org/sites/JCGM/VIM/JCGM_200e_FILES/MAIN_JCGM_200e/01_e.html#L_1_8
and
http://docs.astropy.org/en/stable/units/standard_units.html
But the second option, although less explicit, looks more in line with the description of other quantities which are either dimensionless (such as percentage) or just have one dimension allowed, such as IBU.

Of course, we would use "oneOf"  them as a required object. 

But we can write "amount_as_volume", "amount_as_mass", "amount_as_number", it is probably more explicit. Do you prefer so?

**Comment by mrichar1** [comment]
What is the benefit of using `mass` and `volume` or similar tags, versus just using `amount`? The unit will imply a dimension, and for some SI units they are equivalent (e.g. 1ml water @ room temp = 1g).

**Comment by rmsrosa** [comment]
@mrichar1, as @krutilin and others mentioned, one benefit is readability. I usually keep my recipes in Dropbox, so It sometimes happens to me, for instance. that I want to check a recipe on my mobile or in someone else's computer and don't have access to the app, so readability helps, here. 

It also helps when parsing the json file, since instead of checking the unit against all possible units, and maybe doing some extra conversions, the software first identifies easily what quantity it is and then checks the unit only against the units for that quantity.

Turning the question around, I don't see much benefit in not being explicit.

**Comment by mrichar1** [comment]
I'm not sure it aids readability or parsing that much - you now have to look for 2 potential tags, not 1. You also have a more complex schema - you have to define that it has either mass or volume, but not both (or maybe you *do* allow both? In which case validation becomes tricky). You also have to be sure that someone doesn't get confused and put litres in a mass section, or grams in a volume one - more validation required.

I'm also not 100% sure that every unit can be categorised in terms of mass or volume. There's been some earlier discussion of dealing with uncommon/'unconvertable' units - which occur commonly in brewing. "Squeeze of lemon juice" or "pinch of spice" are valid in recipes, but the 'units' pinch and squeeze would require defining as mass and volumes if the schema mandates it.

I guess the TL:DR; is - what's wrong with the simple option of:

```
"amount": {
                "unit": "XXX",
                "value": N.NN
              }
```

**Comment by krutilin** [comment]
@mrichar1 I proposed this option in the third post of this thread and like it. It is simple and readable enough.

**Comment by rmsrosa** [comment]
@mrichar1 , I meant to allow only one type of amount, by using "oneOf" in the required fields. 

@krutilin, I am confused, you just mentioned 7h ago that you found "not convenient to write units" simply with "amount", without mentioning explicitly what type of quantity it's being used. But if you are okay with @mrichar1 proposition, that would be okay for me too, I just want to get straight the desire of the group and the exact proposition.

As for the complexity and the different routes, let me try to get this right.

(A) One option is to allow one of the descriptions below
```
 "volume": {
                "unit": "ml",
                "value": 8.0
              }
  "mass": {
                "unit": "g",
                "value": 8.0
              }
  "number": {
                "unit": "1",
                "value": 8.0
              }
````
in which case the json schema would have to add these three objects "volume", "mass" and number and have one of them as required (and only one).

(B)  We may allow something like
```
  "amount": {
                "unit": "XXX",
                "value": N.NN
              }
```
where "unit" is not free-form, but only one of the possible units in the quantity types "volume", "mass", and, say,  "number". This will still need the schema to differentiate between "VolumeType", "MassType", and a "NumberType" inside "amount" and allow only one of them, so this is not much less complex and it is a just tiny bit more obscure.

(C) The third option I see is to also have something like
```
  "amount": {
                "unit": "XXX",
                "value": N.NN
              }
```
but not enforce any unit in any case and probably not connect this to measurable_units.json. This would indeed simplify the schema, but I am not sure we want to allow this flexibility, or not to standardize the units, and in my view it will require the codes to be more complex (see below). Is that what you have in mind @mrichar1 ? Or is it option (B) or something else I didn't grasp?

I do see a bit more complexity from the part of the software to parse cases (B) and (C), instead of (A). 

In the case of (C), a software would either need to make sure that every possible volume or mass unit of a given quantity is covered (e.g. including hL - hectoliter - commonly used by some microbreweries) or to hope that other softwares don't come up with a different measure. 

In case (B), the software would need to check for all possible units before parsing to the right object (and assuming it is "free-form" if it is none of the units).

At least in my app, it would be much easier to check for the type of quantity (that would be just three or four options, case (A)) and then parsing the value to that quantity (it is object oriented, of course, so this is seamlessly handled directly by the object, according to its own structure) , instead of checking for all possible units among different objects or against an extra table of units connecting the unit to the right object. 

It is true that I could automatically generate the table just when the software is run, or even just once for each release of the software, taking care not to have the same letter being used for more than one quantity, such as Linter, Lovibond and Liter, or just not including in the table those quantities which are not amounts. It doesn't seem as natural, but it is in fact not a big deal, so I would be okay with (B), too.

But with (C), if we allow free forms like "a squeeze of lemon", possibly handled by one software, the other softwares should be able to understand that and, preferably, how to "squeeze" that information in the few columns usually dedicated to display the unit abbreviations. My solution to that, if we use (A) or (B),  would be to have an ingredient named "squeeze of key lime", or "key lime (squeeze of)" with ``` "amount": { "unit" : "1", "value", 1.0 }``` 

Currently, I am for (A), but I am okay with (B), too. As for (C), it still looks a bit loose and fuzzy to me.

Sorry for the long comment.

**Comment by mrichar1** [comment]
My vote is for (C) as you describe there - this thread is getting a bit long, but I explained why I favour that approach in more detail here: https://github.com/beerjson/beerjson/issues/16#issuecomment-352017853

**Comment by krutilin** [comment]
@rmsrosa no, not units itself but  "volume", "mass", "density" and etc additional descriptions, because it is additional information binds only to the beerjson format. I am ok with units\value constructions and "unit": "ml", "unit": "kg" or "unit": "oz", because I have these values on the packaging of hops or yeasts.

Not sure about free form definitions, so my vote is B. It is a reasonable compromise with complete information.

**Comment by pricelessbrewing** [comment]
Specific heat, while not the SI standard, this is what is used in all brewing software as far as I know.

Cal / (g * deg C) or Cal (g * deg K)

specific volume, used for mash thickness, and grain/hop absorption rates. 

qt/lb, L/kg, gal/lb, fl oz / oz

remove m^3/kg and ft^3/lb. While it is the standard in the scientific world, I'm not aware of any software or recipe that uses cubic meters of ft/lb. 

Going to commit changes to the PR #69 pull request for specific volume and heat units above, but the following needs a discussion. 



>floz vs ozfl

Currently it's listed as ozfl which is non-standard. 

Additionally I think that moving to a single line of unit: "X" makes parsing and reading lovibond/liters/degree linters more difficult. We should switch to an abbreviation. I propose "Liter", " °Lintner", and "°Lovi". Lovi is pretty commonly used as a shorthand for lovibond, so not too strange. Assuming ° symbol is supported. If not, then just drop the °. Thoughts?

**Comment by rmsrosa** [comment]
I didn't think much about the units for specific heat and specific volume, I just added those to have something nonempty for the moment.  And I thought about including Cal / g K, but K is not yet a temperature unit, so I didn't worry much about those units for this commit.  But certainly this should be addressed along the lines that you propose. I also notice that ozfl should be changed to the standard "fl oz" and likewise for other things. This deserves a separate PR.

Concerning Liter, Lintner and Lovibond, they are not mixed together. Lintner and Lovibond to not enter in any "amount" object, so there is no chance for any confusion. I think we should stick with the standard "l" for liter. As for Lintner and Lovibond, they appear in different contexts, as well. We can change Lovibond to Lovi or Lov just to make it shorter and to go along with EBC and SRM. I don't like much the idea of using  ° .

**Comment by pricelessbrewing** [comment]
Density use is inconsistent. Gravity uses type: number in some places, and some places uses DensityUnitType (should change to SpecificGravityUnitType/SpecificDensityUnitType/ExtractUnitType).

**Comment by rmsrosa** [comment]
Hi, all,

It's been a while that I don't contribute. I've got busy with the start of the semester. But there was one thing that I always wanted to do and which is precisely what @pricelessbrewing has just mentioned. I had started to work on it but didn't finish it. This week I managed to update my master branch and create a branch to address these required changes in units. 

I modified `DensityUnitType` to really mean Density and defined `SpecificGravityUnitType` and `ExtractUnitType`. Did the corresponding changes in fermentation, accepting `oneOf` this two forms where appropriate.

In `json/json_culture.json`, I changed `addition_time` and `addition_sg` to `addition_moment`, accepting `oneOf` `TimeType`, `SpecificGravityRangeType`, and `ExtractRangeType`.

In `json/fermentable.json`, I changed `potential` to accept `PercentType` (used by Weyermann, for istance), a newly defined `PointsPerGravityType` (used in the US), and `SpecificGravityType` (which is not quite correct as a unit for potential extract/gravity but which is typically used assuming it means specific gravity per pound per gallon, which is exactly `PointsPerGravity` divided by 1000.

The problem is that I couldn't get the changes to pass the npm test because of that "docs should match snapshot" stuff. I don't understand how this work and what the point is with this. I tried to play with it but it didn't work. Can someone explain it to me? 

If someone wants to see the compare the changes I made, they are in my repository, in the density-extract-units branch: [https://github.com/beerjson/beerjson/compare/master...rmsrosa:density-extract-units](url)

**Comment by pricelessbrewing** [comment]
@rmsrosa Link is broken. 

I'm looking at your repo https://github.com/rmsrosa/beerjson and I don't see a specificgravityrange type or specific gravity unit type? Are you sure you synced to your repo?

>DensityUnitType means actual density

Sounds good. 

>SpecificGravityUnitType vs ExtractUnitType

Hmm not sure why both is necessary. I say we pick one and stick to that. I'm fine with extract or even sugar, with options of sg, plato, brix, balling. 

>Potential

Don't think specificgravityType is quite appropriate for a grain potential. PointsPerGravity or PPG, and percent are the industry standards as far as I know. I would suggest something like 

Potential:
OneOf {
Fine Grind : Percent type
Coarse Grind : PercentType
PPG: PointsPerPoundPerGallonType
Yield: Percent Type
}

Notes for documentation: 

PointsPerPound Per Gallon is defined as the amount of gravity points (Grav = (sg -1) * 1000 ) per lb of fermentable, per gallon of wort. Not per lb per water, but the resultant mixture of sugar solution so to speak. 

Coarse grind, also known as "Yield" and fine grind is defined as the ratio of sugar contributed by the fermentable ingredient, in relation to pure succrose. Succrose has a PPG potential of 46. So for example an ingredient with a potential of 36 PPG, would have a yield (or coarse grind extract potential) of 36/46 =  78.26 %, likely reported at 78.3% +- 0.5% or something like that. Fine grind will typically be 1-3% higher than coarse grind, and is evaluated from mashing with a very very small grind.

**Comment by rmsrosa** [comment]
Sorry for the broken link. I edited and fixed it.

> > SpecificGravityUnitType vs ExtractUnitType
> 
> Hmm not sure why both is necessary. I say we pick one and stick to that. I'm fine with extract or even sugar, with options of sg, plato, brix, balling.

I just wanted to be dimensionally consistent and not mix what are really different units although in practice they are used interchangeably. They are what astropy.units calls equivalencies. Since we are using oneOf, in practice it makes no difference, but I find it more elegant to keep them really as different units.

What you propose for Potential is more complete and I agree with it.

I agree that we shouldn't restrict the addition_moment to "oneOf", but shouldn't we allow for a comment to say "whatever comes first" or should we assume that is implicit? Not sure yet how to handle that or just let it go without "oneOf". 

> Typo in measurable_units.json
> "$ref": "#/definitions/ExtraUnitType" should be "$ref": "#/definitions/ExtractUnitType"

Thanks, I will fix it and check again next week.

**Comment by pricelessbrewing** [comment]
Sounds good, thanks for pointing me towards astropy.units as well. If that's common practice, I'm on board. 

>Potential

Looking at fermentable.json there's already an object property  called yield_dry_basis, potential/PPG belongs in there with the associated units.


```

            "yield_dry_basis": {
              "type": "object",
              "additionalProperties": false,
              "properties": {
                "fine_grind": {
                  "$ref": "measureable_units.json#/definitions/PercentType"
                },
                "coarse_grind": {
                  "$ref": "measureable_units.json#/definitions/PercentType"
                },
                "fine_coarse_difference": {
                  "$ref": "measureable_units.json#/definitions/PercentType"
                }
              },
              "required": ["fine_grind", "coarse_grind"]
},
```

>Addition_moment

I would leave this off of this PR in any case, and move that discussion to #44 since it relates to timing of the culture addition.

**Comment by rmsrosa** [comment]
I fixed the typo in 'measurable units.json', but it still failed the test. The changes in 'package-lock.json' and 'package.json' are due to the update I did in npm and I think it shouldn't be a problem. The problem seems to be when the test suit compares the documentation with the snapshots, but I don't quite understand this process. I am not sure I should edit the documentation myself or the automatic generation of the documentation is already in place somehow.

**Comment by pricelessbrewing** [comment]
@rmsrosa Are you still working on your branch of this? If not I plan on going through and addresses the missing units, and correct gravity vs specific gravity etc. 

Summary of proposed changes:

- Stick to "unit" -"value" pairings. 
- AcidityUnitType : pH. COMPLETE
- Density should refer to Lb/gal, kg/Liter etc. 
- Create specific gravity, plato, extract unit type thing*
- Concentration (ppb, ppm etc) COMPLETE.
- Create length. It was originally there I believe, but I can't find it now, so should be able to go back and just copy paste it. 
- Increase contrast between lintner, lovibond, liter. (lovi, lint, liter)?
- Add Windisch–Kolbach units to diastatic power.
- Specific heat, specific volume. COMPLETE


I still can't think of a good way to combine extract (sugar mass in wort/beer), specific gravity, and plato/balling into one property. In reality they're equivalencies to each other, but technically they are measuring different things. 

Extract is a mass, sg is a unitless dimension of the ratio of density of wort to the density of water, plato is a unitless dimension as a percent of the mass of sugar to the mass of water in wort/beer etc. 

Measurable units is pretty easy. They're seperate unit types. Extract is just a type of mass, so you don't add that as a measurable unit. Create specific gravity, and plato/balling etc are just dimensionless percentages, so they just use percent type. 

The conflict I'm having is what do you call the overall property? Sugar amount? potential is wrong, extract, gravity, and density all feel strange.

**Comment by rmsrosa** [comment]
@pricelessbrewing, no, I'm not working on it. Feel free to address these unit changes. 

My changes didn't pass the npm tests because of the documentation stuff and at some point I lost track of the upstream changes and gave up. But I hope to be able to contribute again in the future. 

As far as I know, extract is not a mass, but it is exactly the ratio of sugar mass to solution or wort mass, and Plato is the percent unit for extract. In my changes, I fixed the definition of density, and introduced SG and Extract (I am not sure whether I introduced gravity points or gravity units, as well, for which 1.040 sg = 40 gp ~ 10 Plato, but it is a good idea to include it). I also introduced ppg (gravity points per pound per gallon, but I erroneously called points per gravity), which is used for fermentable potential.

And I understand the conflict. I don't have a good solution. What I did in my PR was just call it gravity (be it original, apparent, or real) and stick under it an "oneOf" property of one of the above units.
